bigflow/build/pip.py

```diff
@@ -169,3 +169,71 @@ def read_requirements(requirements_path: Path, recompile_check=True) -> List[str
     return result
 
 
+def generate_pinfile(
```

Comment:
This function is long and not very descriptive. There are AT LEAST 2 parts worth implementing as separate functions:

* find conflicts
* resolve conflicts

Reply:
I'm not sure that I understand what you mean by saying "find conflicts" and "resolve conflicts".

This functions is *not* trying to find any conflicts - it receives "find conflicts" as a callback (3rd argument). This is needed because this code must be executed with empty "pinfile" and freshly recompiled "requirements.txt".
Such logic (actually 5 lines of linear code) may be extracted into something like `execute_with_empty_pinfile` (and callback converted to just a list of pins/strings), but I don't think it makes much sense. Did you mean that kind of refactoring?

I've moved `for` loop into a separate function `_try_incrementally_add_pins` -- do you mean this part while saying "resolve conflicts" ? 

Also I've extracted `_include_pinsfile_into_requirements` into a separate function.

Reply:
"I've moved for loop into a separate function _try_incrementally_add_pins -- do you mean this part while saying "resolve conflicts" ?" – yes